# COmapper

**COmapper: High-resolution mapping of meiotic crossovers by long-read sequencing in _Arabidopsis thaliana_**

The project was initiated by Prof. Kyuha Choi, and the code was written by Dohwan Byun at the Plant Genomic Recombination (PGR) Laboratory of Pohang University of Science and Technology (POSTECH).

This research was conducted in collaboration with the following individuals: Dohwan Byun¹, Namil Son¹, Heejin Kim¹, Jaeil Kim¹, Jihye Park¹, Sang-jun Park¹, Hyein Kim¹, Seula Lee², Youbong Hyun², Piotr A. Ziolkowski³, Ian R. Henderson⁴, and Kyuha Choi¹†.

¹Department of Biological Sciences, Pohang University of Science and Technology, Pohang, Gyeongbuk, Republic of Korea

²School of Biological Sciences, Seoul National University, Seoul, Republic of Korea

³Laboratory of Genome Biology, Institute of Molecular Biology and Biotechnology, Adam Mickiewicz University, Poznań, Poland

⁴Department of Plant Sciences, University of Cambridge, Cambridge, UK

The manuscript is available at

# Program version information:
    
COmapper_data_process_ver5_git_upload.sh (2023.09.07 backbone, 2024.08.12 final edit)

COmapper_ver1.2.5_git_upload.py (2024.04.12 backbone, 2024.08.12 final edit)

# System requirement

linux machine with more than 2Tb storage space, 6-core cpu, 32Gb ram

# Required bioinformatics programs
    
    minimap2 https://github.com/lh3/minimap2
    samtools https://www.htslib.org/
    sambamba https://lomereiter.github.io/sambamba/

# Required files (on resources folder)
    
    masksam.awk (script for basefiltering)
    
    TAIR10.mmi (minimap index file) provided as Google Drive link, can also be generated by following steps.

    coller_marker_v4.6.tsv (SNP list file)

# minimap index file setup
    
Download genome sequence

TAIR10 sequence available at https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000001735.3/
    
create index file

    $ minimap2 -d TAIR10.mmi TAIR10.fasta

# Conda environment setup

Create conda environment by following steps

    $ conda create –n COmapper
    $ conda activate COmapper
    $ conda install python=3.7.13
    $ conda install pandas=1.1.3
    $ conda install typing-extensions=4.7.1

# Input data
Nanopore sequencing reads with .fq.gz (or .fq) format

Test input file were provided (fastq_sample.fq)

# Working procedure

1.	Download Nanopore sequencing results as fq.gz file format in the raw data storage folder.

2.	Change User defined arguments of “COmapper_data_process_ver5_git_upload.sh” file
    workdir: folder location to make output files, ensure free space more than 2Tb (~200Gb fq.gz file standard)
  	corenum: number of threads to use
  	name: name of dataset
  	org: location of original fq.gz file

3. Check resource files
   awkloc: location of masksam.awk file
   ref: location of reference genome mmi file (TAIR10.mmi)

4. Save and run COmapper_data_process_ver5.sh.

5. After running COmapper_data_process_ver5.sh, check the output folder for a .tsv file for COmapper analysis.

6. Activate conda environment for Python script.

        $ conda activate COmapper

7. COmapper_ver1.2.5_git_upload.py requires 4 options

   --inputfolder (-i) : input folder location (which includes .tsv files), default: current location
   
   --snpfile (-s) : snp file location, default: current location/collerF2.masked.tiger.txt

   --threads (-t) : thread number, default: 6

   --output (-o) : output file name, default: result.csv

   Set options via command line arguments or manually write on the python code and run COmapper_ver1.2.5_git_upload.py

       $ python COmapper_ver1.2.5_git_upload.py -i INPUTFOLDER -s SNPFILE -t THREADS -o OUTPUTFILE

8. COmapper will return genotyping and classification information via standard output and crossover molecule information in a .csv file.
